### DISCUSSION TITLE
upgraded hard drives - what to do with u-boot?

### DISCUSSION THREAD

**Discussion by ringprince** [1000001101000/Debian_on_Buffalo#127] [Q&A] [answered]
Hi, 

I've been running debian (buster) on my LS-WXL successfully after using your installation script.  Thanks a lot for this!!!

My previous setup was built on two 2TB hard drives in RAID (RAID0 for swap, RAID1 for /, /boot, and /home).

I've now upgraded the hard disks rather naively to two 14TB hard drives.  I did not think properly when doing this, as I have prepared the hard drives with UEFI boot partition and hoping to just run grub-install at the end.

Now, I realized of course, that there is no grub but u-boot.  I do not know much (as in 'anything') about u-boot and I am here to ask for help before I restart the system ;-)

I assume that I have to do something, like re-generating some image?
Will I run into problems as my new drives are gpt-partions now?
Will I run into problems as my /home partition is in RAID with metadata version 1.2 ?

Any help is highly appreciated!

Kind regards,
ringprince

For starters, here is the outpout of `cat /proc/mdstat` (I am of course happy to produce more information):

```
Personalities : [raid0] [raid1] [linear] [multipath] [raid6] [raid5] [raid4] [raid10]
md93 : active raid0 sdc3[1] sdb3[0]
      16758784 blocks super 1.2 512k chunks

md91 : inactive sda2[1]
      8388544 blocks

md127 : active raid1 sdb5[3] sdc5[2]
      1910649856 blocks super 1.2 [2/2] [UU]
      bitmap: 0/15 pages [0KB], 65536KB chunk

md90 : active raid1 sdb2[0] sdc2[1]
      786368 blocks [2/2] [UU]

md92 : active raid1 sdb4[0] sdc4[1]
      33555392 blocks [2/2] [UU]

unused devices: <none>
```

(In case you wonder about the inactive /dev/md91: I had trouble re-building the RAID0 volume and, thus, created a new one)

EDIT:
I forgot to mention, that I have updated `/etc/fstab` and `/etc/mdadm/mdadm.conf` and I have run ` update-initramfs -u`

**Comment by 1000001101000** [comment]
Good to hear from you!

I'm not 100% what will or won't work but this sounds like an interesting experiment. 

The only hard limit I'm aware of is the 16TB (TiB?) volume limit which comes from this architecture using 4k pages/blocks on a 32-bit architecture. I believe a 14TB RAID1 should be fine though. There could be other problems, I doubt buffalo had any 14TB drives to test with when they made these devices but I haven't run into many of those yet.......heck I've had more problems with newer PC motherboards and sata drives than I ever have with buffalo devices. 

You might be able to run through the normal install process but I'm not certain. Since the drives are larger than 2TB they will eventually need a GPT partition table (assuming you plan to use more than 2TB of the drive). I'm fairly sure all models newer than that one can handle that without any special considerations, I think that one can too. 

I would think if you create a suitable /boot on a drive (partition#1, 1GB, ext3, plain or raid1 with meta 0.9) with the installer files the device could probably boot it. I'd try first with a GPT if possible then try MBR if it seems to fail booting. 

The metadata=0.90 limitation only applies to /boot, metadata 1.2 should be fine for the other arrays. 

If I'm wrong about this device booting GPT you can still probably make it work by switching to a Hybrid GPT/MBR after getting up and running with just an MBR and converting. 

I have a script that can be used to generate disk images rather than the installer which can make this process a little easier for uncooperative devices. More importantly the documentation includes information on converting to hybrid MBR so you can use pretty much any disk on even the oldest devices. 
https://github.com/1000001101000/Debian_on_Buffalo/wiki/Alternate-install-method-via-debootstrap-script

Let me know how it goes!

**Comment by ringprince** [comment]
Great, thanks for this fast reaction and detailed answer.

Just to be sure, I understand correctly:  You are suggesting that I install again from scratch?

Since the device is running right now (and all limits you mention seem to be fulfillde, like boot is on metadata=0.90, no volume>16TB) I'd naturally prefer to just reboot.
If there are no obvious things I forgot to do, I would go ahead and try that.

**Comment by 1000001101000** [comment]
sorry, I guess I'm a little unclear what state you have things in. I actually assumed the information posted above was from the drive hooked up to a PC or something. 

could you post the output of `lsblk` from the device?

**Comment by ringprince** [comment]
sure:

```
NAME      MAJ:MIN RM  SIZE RO TYPE  MOUNTPOINT
sdb         8:16   0 12.8T  0 disk
|-sdb1      8:17   0    1M  0 part
|-sdb2      8:18   0  768M  0 part
| `-md90    9:90   0  768M  0 raid1 /boot
|-sdb3      8:19   0    8G  0 part
| `-md93    9:93   0   16G  0 raid0 [SWAP]
|-sdb4      8:20   0   32G  0 part
| `-md92    9:92   0   32G  0 raid1 /
`-sdb5      8:21   0 12.7T  0 part
  `-md127   9:127  0  1.8T  0 raid1 /home
sdc         8:32   0 12.8T  0 disk
|-sdc1      8:33   0    1M  0 part
|-sdc2      8:34   0  768M  0 part
| `-md90    9:90   0  768M  0 raid1 /boot
|-sdc3      8:35   0    8G  0 part
| `-md93    9:93   0   16G  0 raid0 [SWAP]
|-sdc4      8:36   0   32G  0 part
| `-md92    9:92   0   32G  0 raid1 /
`-sdc5      8:37   0 12.7T  0 part
  `-md127   9:127  0  1.8T  0 raid1 /home
mtdblock0  31:0    0  384K  1 disk
mtdblock1  31:1    0   64K  1 disk
mtdblock2  31:2    0   64K  0 disk
```

EDIT: haven't re-grown the home partition yet...

**Comment by 1000001101000** [comment, marked as answer]
ah, okay. i don't think that will boot as-is but I think you can tweak it so that it would.

I believe you need to remove those tiny EFI partitions and then renumber your partitions so that both /boot are partition number 1, uboot only looks at partition 1. I know that having a partition 2 at the beginning of the disk will be ignored, I'm not sure what happens if there are partitions before partition 1 so I'd just delete those and renumber. 

Also run `update-initramfs -u` before trying to reboot.

**Comment by ringprince** [comment]
Thanks a lot, good to know.  I'll do that and then try.  (Will take a while on the running system with disabling and re-assembling the RAID, though).  I'll report back.

**Comment by 1000001101000** [comment]
I wouldn't think you'd need to mess with the RAID, though admittedly it's been a while since I re-numbered a partition

**Comment by 1000001101000** [comment]
since it's just /boot you could probably unmount and then stop the raid, re-number the partitions then start and mount. If it doesn't like something you can probably force assemble and update mdadm.conf.

**Comment by ringprince** [comment]
Reboot went fine!  Thanks a ton.  I would not have figured that out easily without your help regarding the first partition!
